### PR TITLE
[Form] issue-13589: adding custom false-values to BooleanToString transformer

### PR DESCRIPTION
**Pull request by leberknecht**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #13589
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/9031

As suggested in #13589 , this PR adds the option to specify custom false-values, so we can use the CheckBoxType to handle strings '1' / '0' and eval them to true / false. While HTTP defines that checkbox types are either submitted=true or not-submitted=false, no matter of what value was submitted, it would be nice to have this option. 
Also refs (which read like "the basic idea of the feature was accepted, PR almost done, then something-happend so PR wasnt merged"..?)

https://github.com/symfony/symfony/pull/15054
https://github.com/symfony/symfony/pull/18005

### CODE REVIEW

File: src/Symfony/Component/Form/Extension/Core/Type/CheckboxType.php [hunk 1]
```diff
@@ -59,6 +59,7 @@ public function configureOptions(OptionsResolver $resolver)
             'value' => '1',
             'empty_data' => $emptyData,
             'compound' => false,
+            'false_values' => array(null),
```
Comment (xabbuh) [inline src/Symfony/Component/Form/Extension/Core/Type/CheckboxType.php]: I think we should also force the option to be an array to not break the type hint of the transformer's constructor

Comment (leberknecht) [reply src/Symfony/Component/Form/Extension/Core/Type/CheckboxType.php]: fixed

File: src/Symfony/Component/Config/Tests/Util/XmlUtilsTest.php [hunk 2]
```diff
@@ -55,7 +55,7 @@ public function testLoadFile()
             XmlUtils::loadFile($fixtures.'valid.xml', array($mock, 'validate'));
             $this->fail();
         } catch (\InvalidArgumentException $e) {
-            $this->assertRegExp('/The XML file "[\w:\/\\\.]+" is not valid\./', $e->getMessage());
+            $this->assertRegExp('/The XML file "[\w:\/\\\.-]+" is not valid\./', $e->getMessage());
```
Comment (xabbuh) [inline src/Symfony/Component/Config/Tests/Util/XmlUtilsTest.php]: Where is this change coming from?

File: src/Symfony/Component/Form/Tests/Extension/Core/Type/CheckboxTypeTest.php [hunk 3]
```diff
@@ -173,6 +174,36 @@ public function provideCustomModelTransformerData()
...
+        return array(
+            array(''),
+            array('false'),
+            array('0'),
+        );
+    }
+
+    public function testDontAllowNonArrayFalseValues()
+    {
+        $this->expectException(InvalidOptionsException::class);
```
Comment (xabbuh) [inline src/Symfony/Component/Form/Tests/Extension/Core/Type/CheckboxTypeTest.php]: I think we use the `@expectedException` annotation everywhere instead

File: src/Symfony/Component/Form/CHANGELOG.md [hunk 4]
```diff
@@ -19,6 +19,7 @@ CHANGELOG
  * removed passing guesser services ids as the fourth argument of `DependencyInjectionExtension::__construct()`
  * removed the ability to validate an unsubmitted form.
  * removed `ChoiceLoaderInterface` implementation in `TimezoneType`
+ * added support for custom false-values to BooleanToStringTransformer and CheckBoxType, so e.g. submitted string "false" will be evaluated to false instead of true
```
Comment (xabbuh) [inline src/Symfony/Component/Form/CHANGELOG.md]: I would reword this a bit:

```markdown
 * added the `false_values` option to the `CheckboxType` which allows to configure custom values which
   will be treated as `false` during submission
```

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/BooleanToStringTransformer.php [hunk 5]
```diff
@@ -24,12 +24,16 @@ class BooleanToStringTransformer implements DataTransformerInterface
...
     /**
-     * @param string $trueValue The value emitted upon transform if the input is true
+     * @param string $trueValue   The value emitted upon transform if the input is true
+     * @param array  $falseValues
      */
-    public function __construct(string $trueValue)
+    public function __construct(string $trueValue, array $falseValues = array(null))
     {
         $this->trueValue = $trueValue;
+        $this->falseValues = $falseValues;
```
Comment (xabbuh) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/BooleanToStringTransformer.php]: IMO we should also add some error handling in case `$trueValue` is one of the elements in `$falseValues`

File: src/Symfony/Component/Config/Tests/Util/XmlUtilsTest.php (hunk 2 above)
Comment (leberknecht) [reply src/Symfony/Component/Config/Tests/Util/XmlUtilsTest.php]: uh, that is actually unrelated to the feature, i can revert it (my tests were failing as i had a dash in my cwd pathname). Should i move it out?

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/BooleanToStringTransformer.php (hunk 5 above)
Comment (leberknecht) [reply src/Symfony/Component/Form/Extension/Core/DataTransformer/BooleanToStringTransformer.php]: Is InvalidArgumentException ok?

File: src/Symfony/Component/Config/Tests/Util/XmlUtilsTest.php (hunk 2 above)
Comment (xabbuh) [reply src/Symfony/Component/Config/Tests/Util/XmlUtilsTest.php]: yes please, but we should take a look if we should make this change in the `2.7` branch instead (in a separate PR)

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/BooleanToStringTransformer.php (hunk 5 above)
Comment (xabbuh) [reply src/Symfony/Component/Form/Extension/Core/DataTransformer/BooleanToStringTransformer.php]: yes

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/BooleanToStringTransformer.php [hunk 6]
```diff
@@ -24,12 +25,19 @@ class BooleanToStringTransformer implements DataTransformerInterface
...
+        if (in_array($this->trueValue, $this->falseValues)) {
```
Comment (xabbuh) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/BooleanToStringTransformer.php]: please pass `true` as the third argument here

**Review by xabbuh — APPROVED** [review]

**Comment by leberknecht** [comment]
:) Thanks for your help!

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @leberknecht.
